### DISCUSSION TITLE
🆕 Welcome to the code search and navigation beta!

### DISCUSSION THREAD

**Discussion by look** [community/community#38734] [Other Feature Feedback, Questions, & Ideas]
We are pleased to announce that we have integrated all the features of cs.github.com -- and more -- directly into GitHub.com with the [new code search and code view experience](https://github.com/features/code-search). 

![hero-search-popover](https://user-images.githubusercontent.com/10186/200931959-7647a534-5f2c-467a-ba9a-eb590db33357.png)

This release consists of three big new features:

1. A redesigned search experience with syntax highlighting, autocomplete, and powerful search capabilities
2. A custom-built code search engine that supports exact match and regular expressions
3. The new code view, with an integrated file tree, symbol view, and code navigation

Everyone who signed up for the Code Search Technology Preview has been added to the beta[^1]. If you didn't get access, [please sign up](https://github.com/features/code-search-code-view/signup). We clear the waitlist frequently.

Thank you for all your feedback about the Code Search Technology Preview. It's been invaluable as we built this functionality into GitHub.com, so keep it coming!

[^1]: To switch back to the old search and file view experience, click the Feature preview option under your profile.

**Comment by JoeZiminski** [comment]
This looks great, is it possible to search within specific branches?

**Reply by look** [reply]
Not yet! This is a feature we plan to add, we designed the new system to support it.

**Comment by JoeZiminski** [comment]
Thanks, great to hear!

**Comment by tonglil** [comment]
`ctrl-f` feedback here: https://github.com/orgs/community/discussions/39163#discussioncomment-4180000

**Comment by eddumelendez** [comment]
is there an API to play with?

**Reply by look** [reply]
An API is on our roadmap. Can you share some of your use cases for an API, and the kind of volume you are looking for?

**Reply by eddumelendez** [reply]
maybe shouldn't duplicate the comment but it was answered [here](https://github.com/orgs/community/discussions/38271#discussioncomment-4061038)

Looking forward to it.

**Reply by ng-druid** [reply]
It would be nice to filter/search JSON using JSONPath query as part of new search syntax.

**Reply by Ollie-Armitage** [reply]
My current use case where it could prove valuable is I want to be able to compound and intersect two searches as my requirements are slightly complicated - Specifically with the release of NextJS13, I want to be able to search for repos that have an 'app' directory (a new optional feature in NextJS 13), and separately I want to be able to search for repos that have a package.json specifying the Next version. If I have those two I can then find the set of repos contained in both myself.

**Comment by ng-druid** [comment]
Frequently when searching others libraries I'm left to my own devices to retrace the path. It would be cool for search ui to introduce something like a "Code Journey". A code journey would be a session that once begins records the journey as you step through various files and functions. Developers could either explicitly add a journey item when moving between functions/files or have the journey auto record everything. So even though you may have stepped though several files and functions to find an answer the steps can be retraced back to the origin via journey recordings. Devs could also save journeys to come back to later or refresh their memory about how something works in a specific repository.

**Comment by JulianLeRoux** [comment]
Firstly, the new code search and code view are _amazing_. Kudos to the team.

I think this may be unexpected behavior from the user perspective:
1. Search using`cs.github.com`, where the search term is scoped to your organization
2. In the results page, click the link `New! Try this search in GitHub.com`
3. The corresponding Github.com results page is for the query `saved:<organization_name> <foo>`, rather than the expected `org:<organization_name> <foo>`.
4. Sending the URL of these query results to another person will return nothing for them, presumably because they don't have the same saved scopes.

Further, I could not find any documentation for the `saved` qualifier at either
[understanding-github-code-search-syntax](https://docs.github.com/en/search-github/github-code-search/understanding-github-code-search-syntax) or [Search syntax reference](https://cs.github.com/about/syntax)

Thanks.

**Comment by mpoindexter** [comment]
I think the new code search and code view are absolutely fantastic! As an Enterprise Cloud organization member, I wish there was an easy/good way to see usages of a symbol within my organization. Other than that, I think this is a giant leap forward and a wonderful feature.

**Comment by SampsonCrowley** [comment]
While initially fantastic at finding code usages, it seems that the search has lost symbols recently...

I recently did a search on cs.github.com and only found 2 matches for a _very simple_ class name that is used many times in the organization across multiple repos.

pulling the code down and searching with my editor found many many more usages than what was found by `cs.github.com` where initially after launch this feature used to find everything I was looking for

**Reply by SampsonCrowley** [reply]
1 example is simply searching for the word `Ordering` on `rails/arel`

the only result is in the HISTORY.txt file, despite it very clearly being used in multiple files, and defined in `lib/arel/nodes/unary.rb`

**Reply by look** [reply]
@SampsonCrowley it looks like rails/arel no longer has any code in it, and the code has been moved into rails/rails. If you search there, there are [many results](https://github.com/search?q=repo%3Arails%2Frails%20Ordering&type=code).

**Reply by SampsonCrowley** [reply]
> @SampsonCrowley it looks like rails/arel no longer has any code in it, and the code has been moved into rails/rails. If you search there, there are [many results](https://github.com/search?q=repo%3Arails%2Frails%20Ordering&type=code).

No, there is still plenty of in 6 different release branches. If code is only possible to search in the default branch, that is a major home in ts capacities

**Reply by look** [reply]
Currently, code search only indexes the default branch of a repository. We plan to add branch search, but are focusing on scaling the system first. There is [an older discussion about branch search](https://github.com/orgs/community/discussions/8564) you can check out and contribute to.

**Comment by Aeolun** [comment]
It seems like using an @ symbol in an exact match search for npm package names breaks the search results? If I use "@org/packagename" I have results missing, but if I search for "org/packagename" they show up.

**Comment by wborn** [comment]
Do you know if this beta has caused code search results to break for certain repositories?

Users that are not using the beta can no longer search code in https://github.com/openhab/openhab-addons

See also this discussion: https://github.com/orgs/community/discussions/44388

**Comment by gipnomag** [comment]
> We are pleased to announce that we have integrated all the features of cs.github.com -- and more -- directly into GitHub.com with the [new code search and code view experience](https://github.com/features/code-search).
> 
> ![hero-search-popover](https://user-images.githubusercontent.com/10186/200931959-7647a534-5f2c-467a-ba9a-eb590db33357.png)
> 
> This release consists of three big new features:
> 
> 1. A redesigned search experience with syntax highlighting, autocomplete, and powerful search capabilities
> 2. A custom-built code search engine that supports exact match and regular expressions
> 3. The new code view, with an integrated file tree, symbol view, and code navigation
> 
> Everyone who signed up for the Code Search Technology Preview has been added to the beta[1](#user-content-fn-1-14ab0865dc8bb9945613783a06bffbc2). If you didn't get access, [please sign up](https://github.com/features/code-search-code-view/signup). We clear the waitlist frequently.
> 
> Thank you for all your feedback about the Code Search Technology Preview. It's been invaluable as we built this functionality into GitHub.com, so keep it coming!
> 
> ## Footnotes
> 1. To switch back to the old search and file view experience, click the Feature preview option under your profile. [↩](#user-content-fnref-1-14ab0865dc8bb9945613783a06bffbc2)
